src/main/drivers/serial_uart_stm32h7xx.c

```diff
@@ -0,0 +1,567 @@
+        .rxPins = {
+            { DEFIO_TAG_E(PE0), GPIO_AF8_UART8 }
+        },
+        .txPins = {
+            { DEFIO_TAG_E(PE1), GPIO_AF8_UART8 }
+        },
+        .rcc_apb1 = RCC_APB1L(UART8),
+        .rxIrq = UART8_IRQn,
+        .txPriority = NVIC_PRIO_SERIALUART8_TXDMA,
+        .rxPriority = NVIC_PRIO_SERIALUART8,
+        .txBuffer = uart8TxBuffer,
+        .rxBuffer = uart8RxBuffer,
+        .txBufferSize = ARRAYLEN(uart8TxBuffer),
+        .rxBufferSize = ARRAYLEN(uart8RxBuffer),
+    },
+#endif
+};
+
+#ifdef USE_DMA
+static void handleUsartTxDma(uartPort_t *s);
+#endif
+
+void uartIrqHandler(uartPort_t *s)
+{
+    UART_HandleTypeDef *huart = &s->Handle;
+    /* UART in mode Receiver ---------------------------------------------------*/
+    if ((__HAL_UART_GET_IT(huart, UART_IT_RXNE) != RESET)) {
+        uint8_t rbyte = (uint8_t)(huart->Instance->RDR & (uint8_t) 0xff);
+
+        if (s->port.rxCallback) {
+            s->port.rxCallback(rbyte, s->port.rxCallbackData);
+        } else {
+            s->port.rxBuffer[s->port.rxBufferHead] = rbyte;
+            s->port.rxBufferHead = (s->port.rxBufferHead + 1) % s->port.rxBufferSize;
+        }
+        CLEAR_BIT(huart->Instance->CR1, (USART_CR1_PEIE));
+
+        /* Disable the UART Error Interrupt: (Frame error, noise error, overrun error) */
+        CLEAR_BIT(huart->Instance->CR3, USART_CR3_EIE);
+
+        __HAL_UART_SEND_REQ(huart, UART_RXDATA_FLUSH_REQUEST);
+    }
+
+    /* UART parity error interrupt occurred -------------------------------------*/
+    if ((__HAL_UART_GET_IT(huart, UART_IT_PE) != RESET)) {
+        __HAL_UART_CLEAR_IT(huart, UART_CLEAR_PEF);
+    }
+
+    /* UART frame error interrupt occurred --------------------------------------*/
+    if ((__HAL_UART_GET_IT(huart, UART_IT_FE) != RESET)) {
+        __HAL_UART_CLEAR_IT(huart, UART_CLEAR_FEF);
+    }
+
+    /* UART noise error interrupt occurred --------------------------------------*/
+    if ((__HAL_UART_GET_IT(huart, UART_IT_NE) != RESET)) {
+        __HAL_UART_CLEAR_IT(huart, UART_CLEAR_NEF);
+    }
+
+    /* UART Over-Run interrupt occurred -----------------------------------------*/
+    if ((__HAL_UART_GET_IT(huart, UART_IT_ORE) != RESET)) {
+        __HAL_UART_CLEAR_IT(huart, UART_CLEAR_OREF);
+    }
+
+    /* UART in mode Transmitter ------------------------------------------------*/
+    if (
+#ifdef USE_DMA
+        !s->txDMAStream &&
+#endif
+        (__HAL_UART_GET_IT(huart, UART_IT_TXE) != RESET)) {
+        /* Check that a Tx process is ongoing */
+        if (huart->gState != HAL_UART_STATE_BUSY_TX) {
+            if (s->port.txBufferTail == s->port.txBufferHead) {
+                huart->TxXferCount = 0;
+                /* Disable the UART Transmit Data Register Empty Interrupt */
+                CLEAR_BIT(huart->Instance->CR1, USART_CR1_TXEIE);
+            } else {
+                if ((huart->Init.WordLength == UART_WORDLENGTH_9B) && (huart->Init.Parity == UART_PARITY_NONE)) {
```

Comment:
Does BF use 9bit mode?

Reply:
I think this is needed to support proper parity checking for SBus with even parity.

Reply:
buffer is 8-bit, this code will take 9 bit from other uart byte

Reply:
This code was (iirc) derived from F7 variant of the code (`serial_uart_stm32f7xx.c`).

The code section under question actually seems to handle 9-bit per character transmission mode (for whatever purpose), and yes, buffer is handled as an array of 16-bit words here.

Since this is the transmitter interrupt, I think the whole section is redundant, but will leave it as is and register it in the H7 code improvement note.

Reply:
`txBuffer` is `uint8_t*` and `txBufferTail` is incremented by one, so this won't work anyway (buffer index `[]` has higher precedence than type cast)

Reply:
Removed the code and it seems working for F.Port rx/tx case.

Reply:
?
Not sure here ... probably fix it eventually with F7 ...